en/lessons/basics/modules.md

```diff
@@ -1,316 +1,199 @@
-- `behaviour` — Use an OTP or user-defined behaviour.
-
-## Structs
+Here is the same module, but residing at a different filepath.
 
-Structs are special maps with a defined set of keys and default values.
-A struct must be defined within a module, which it takes its name from.
-It is common for a struct to be the only thing defined within a module.
+```haskell
+-- src/Data/Examples/Example.hs
+module Data.Examples.Example where
 
-To define a struct we use `defstruct` along with a keyword list of fields and default values:
+id :: a -> a
+id a = a
 
-```elixir
-defmodule Example.User do
-  defstruct name: "Sean", roles: []
-end
+compose :: (b -> c) -> (a -> b) -> (a -> c)
+compose f g = \a -> f (g a)
 ```
 
-Let's create some structs:
+Modules can export a subset of the entities they contain. In the example below
+only the type `Identity` (and its constructor) and `id` are exported. The
+`compose` function is completely isolated to the local module scope.
 
-```elixir
-iex> %Example.User{}
-%Example.User<name: "Sean", roles: [], ...>
+```haskell
+-- src/Identity.hs
+module Identity (
+  Identity(..),
+  id
+) where
 
-iex> %Example.User{name: "Steve"}
-%Example.User<name: "Steve", roles: [], ...>
+data Identity a = Identity a
 
-iex> %Example.User{name: "Steve", roles: [:manager]}
-%Example.User<name: "Steve", roles: [:manager]>
-```
-
-We can update our struct just like we would a map:
+id :: a -> a
+id a = a
 
-```elixir
-iex> steve = %Example.User{name: "Steve"}
-%Example.User<name: "Steve", roles: [...], ...>
-iex> sean = %{steve | name: "Sean"}
-%Example.User<name: "Sean", roles: [...], ...>
+compose :: (b -> c) -> (a -> b) -> (a -> c)
+compose f g = \a -> f (g a)
 ```
+__Note__: you'll learn about using and defining types in Haskell in
+[types](../types) and [algebraic data types](../algebraic-data-types).
 
-Most importantly, you can match structs against maps:
+Modules can also export types, but hide constructors. This means (with respect
+to the example below) that you can speak about things of type `Identity` but you
+cannot construct them yourself. These are sometimes referred to as opaque types.
 
-```elixir
-iex> %{name: "Sean"} = sean
-%Example.User<name: "Sean", roles: [...], ...>
-```
+```haskell
+-- src/Identity.hs
+module Identity (
+  Identity,
+  wrapIdentity
+) where
 
-As of Elixir 1.8 structs include custom introspection.
-To understand what this means and how we are to use it let us inspect our `sean` capture:
+data Identity a = Identity a
 
-```elixir
-iex> inspect(sean)
-"%Example.User<name: \"Sean\", roles: [...], ...>"
+wrapIdentity :: a -> Identity a
+wrapIdentity = Identity
 ```
 
-All of our fields are present which is okay for this example but what if we had a protected field we didn't want to include?
-The new `@derive` feature let's us accomplish just this!
-Let's update our example so `roles` are no longer included in our output:
+There is only one piece of syntax that ever comes before module syntax; that is
```

Comment:
FIY: Actually, module documentation for Haddock also comes before module syntax. It's in the form:

```haskell
{-|
Module      : W
Description : Short description
Copyright   : (c) Some Guy, 2013
                  Someone Else, 2014
License     : GPL-3
Maintainer  : sample@email.com
Stability   : experimental
Portability : POSIX

Here is a longer description of this module, containing some
commentary with @some markup@.
-}
module W where
```

We can link to the haddock lesson when it's made (but that's not for basic track)

Reply:
I would argue that that isn't a piece of syntax, but added a another comment.